### PR TITLE
Allow transparent barrierColor in showDialog methods

### PR DESCRIPTION
**Pull request by tvolkert**
## Description

`showDialog()` and `showGeneralDialog()` were allowing transparent
colors but then triggering an assertion down the stack while building
the modal barrier. The assertion existed to keep from animating from
transparent to transparent, but there's no need for the assertion,
since we can just treat the transparent case as the same as the null
case -- no animation necessary.

## Tests

I added the following tests:

* A test that ensures that you can create a dialog with a transparent barrier color
* A test for barrierDismissable, since such a test didn't exist

## Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*

### CODE REVIEW

File: packages/flutter/test/widgets/routes_test.dart [hunk 1]
```diff
@@ -966,6 +966,84 @@ void main() {
       expect(secondaryAnimationOfRouteOne.value, primaryAnimationOfRouteTwo.value);
     });
 
+    testWidgets('showGeneralDialog handles transparent barrier color', (WidgetTester tester) async {
+      await tester.pumpWidget(MaterialApp(
+        home: Navigator(
```
[the remark this answers is not in the document]
Comment (goderbauer) [reply packages/flutter/test/widgets/routes_test.dart:971]: You'd also end up with less modal barriers making the expect below a little more sensible :)

**Review by goderbauer — COMMENTED** [review]
LGTM after questions on the test are resolved.

File: packages/flutter/test/widgets/routes_test.dart (hunk 1 above)
[the remark this answers is not in the document]
Comment (tvolkert) [reply packages/flutter/test/widgets/routes_test.dart]: Done

**Review by goderbauer — APPROVED** [review]
LGTM

**Pull request merged**